diff --git a/addon/services/asset-loader.js b/addon/services/asset-loader.js
index a741fcfd530c146cd3dc36de1a8693c1306a950e..0c4adacda1e5d6a07f60232779922724ac5e0056 100644
--- a/addon/services/asset-loader.js
+++ b/addon/services/asset-loader.js
@@ -1,5 +1,6 @@
 import RSVP from 'rsvp';
-import Ember from 'ember';
+import Service from '@ember/service';
+import { assert } from '@ember/debug';
 import AssetLoadError from '../errors/asset-load';
 import BundleLoadError from '../errors/bundle-load';
 import JsLoader from '../loaders/js';
@@ -22,7 +23,7 @@ function reduceManifestBundles(input, manifest) {
 
   // Merge the bundles together, checking for collisions
   return Object.keys(manifest.bundles).reduce((output, bundle) => {
-    Ember.assert(`The bundle "${bundle}" already exists.`, !output.bundles[bundle]);
+    assert(`The bundle "${bundle}" already exists.`, !output.bundles[bundle]);
     output.bundles[bundle] = manifest.bundles[bundle];
     return output;
   }, input);
@@ -33,7 +34,7 @@ function reduceManifestBundles(input, manifest) {
  *
  * @class AssetLoader
  */
-export default Ember.Service.extend({
+export default Service.extend({
   /**
    * Setup the caches for the service to use when loading assets.
    *
@@ -171,7 +172,7 @@ export default Ember.Service.extend({
   getManifest() {
     const manifest = this.__manifest;
 
-    Ember.assert('No asset manifest found. Ensure you call pushManifest before attempting to use the AssetLoader.', manifest);
+    assert('No asset manifest found. Ensure you call pushManifest before attempting to use the AssetLoader.', manifest);
 
     return manifest;
   },
@@ -233,11 +234,11 @@ export default Ember.Service.extend({
 
     const bundles = manifest.bundles;
 
-    Ember.assert('Asset manifest does not list any available bundles.', Object.keys(bundles).length);
+    assert('Asset manifest does not list any available bundles.', Object.keys(bundles).length);
 
     const bundle = bundles[name];
 
-    Ember.assert(`No bundle with name "${name}" exists in the asset manifest.`, bundle);
+    assert(`No bundle with name "${name}" exists in the asset manifest.`, bundle);
 
     return bundle;
   },
@@ -253,7 +254,7 @@ export default Ember.Service.extend({
   _getAssetLoader(type) {
     const loader = this.__assetLoaders[type];
 
-    Ember.assert(`No loader for assets of type "${type}" defined.`, loader);
+    assert(`No loader for assets of type "${type}" defined.`, loader);
 
     return loader;
   },
diff --git a/addon-test-support/preload-assets.js b/addon-test-support/preload-assets.js
index 762cddd4efee9e174de85bb6851974e07e5c3681..2f4892b23f4bc9d21be94d5a083f01653cfaac40 100644
--- a/addon-test-support/preload-assets.js
+++ b/addon-test-support/preload-assets.js
@@ -1,14 +1,11 @@
-import Ember from 'ember';
+import { all } from 'rsvp';
 import AssetLoader from 'ember-asset-loader/services/asset-loader';
 
-const { Test, RSVP } = Ember;
-
 /**
  * Preloads all the bundles specified in an asset manifest
  * to make sure all files are available for testing.
  *
- * Uses the Ember.Test.Promise class to make sure tests
- * wait for the assets to load first.
+ * Returns a promise so tests wait for the assets to load first.
  *
  * @return {Promise}
  */
@@ -17,7 +14,5 @@ export default function preloadAssets(manifest) {
   loader.pushManifest(manifest);
 
   const bundlePromises = Object.keys(manifest.bundles).map((bundle) => loader.loadBundle(bundle));
-  const allBundles = RSVP.all(bundlePromises);
-
-  return Test.resolve(allBundles);
+  return all(bundlePromises);
 }
